### PR TITLE
Maljdan/plotting legend

### LINKED ISSUE #122: Plotting Legends in RAVEN
--------
Issue Description
--------
##### What did you expect to see happen?

I expect the matplotlib legend interface to be exposed to the user so they can plot legends on their plots.

##### What did you see instead?

This capability is restricted and not recognized by the raven input.

##### Do you have a suggested fix for the development team?

I had created a branch on gitlab and I am now porting it over to github so it can be incorporated into our main devel branch.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

No file as feature is not yet included.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements? __Does not impact requirements__
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.  __Not a defect__
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [ ] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by maljovec**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

Closes #122 

##### What are the significant changes in functionality due to this change request?

This adds the capability to add a legend to a plot and accepting most of the arguments of [matplotlib's legend function](http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.legend). This also adds brief documentation of this feature and a test that generates a plot with a legend on it.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: doc/user_manual/OutStreamSystem.tex [hunk 1]
```diff
@@ -880,6 +879,11 @@ \subsubsection{Plot input structure \label{sec:PlotInputStructure}}
 \end{itemize}
 %
 
+One may also specify a \xmlNode{legend} tag that will place a legend on the plot.
+The legend accepts the same arguments as the matplotlib interface for creating a
+legend (refer to the \href{http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.legend}{matplotlib API} for details.
+Note, that only primitive datatypes are accepted in the raven input file so values like bbox\_transform, prop, and handler\_map are not available through the raven interface).
+
```
Comment (alfoa) [inline doc/user_manual/OutStreamSystem.tex]: Dan, Can you explicit the "options" available (like for the other "capabilities")? 
Listing the options are available through our interface (the one find-able in http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.legend)

Comment (alfoa) [inline doc/user_manual/OutStreamSystem.tex]: At least the main options...

File: tests/framework/OutStreams/test_legend.xml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" ?>
```
Comment (alfoa) [inline tests/framework/OutStreams/test_legend.xml]: This input file does not work in my machine. I do not get any legend. 
legendtest


>>> import matplotlib
>>> matplotlib.__version__
'1.5.1'

**Comment by maljovec** [comment]
My apologies, my edits did not make it through on the first attempt. The output of test_legend.xml in the `tests/framework/OutStreams/` directory should look like the attached image:
legendtest

If it does not, then we still have a problem. The colors may vary if you are using an older version of matplotlib, but the legend should appear in the bottom center of the plot.

**Comment by alfoa** [comment]
In matplotlib 1.5.1 the legend is at the top right corner. But I think it is due to the different version of matplotlib. 

In addition, you should clearly indicate in the manual that if you want a meaningful legend, you are forced to specify the 
```XML
  <kwargs>
     <label> bla bla bla </label>
  </kwargs>
```
For the one that can accept that (or something else for the plots that do not).

Thanks


legendtest

File: doc/user_manual/OutStreamSystem.tex (hunk 1 above)
Comment (maljovec) [inline doc/user_manual/OutStreamSystem.tex]: I attempted to do this in the latest commit. The result looks like this:

legendtext

Comment (maljovec) [inline doc/user_manual/OutStreamSystem.tex]: I am leaving more advanced features out for now as they are untested, and I think these should be enough to get the user started.

**Comment by maljovec** [comment]
I have added an example and some more text to help illuminate how the legend should be used:

legendexample

**Comment by maljovec** [comment]
I am not sure why these tests are failing: mpiqsub_nosplit, mpi. I am trying to gain access to an HPC node in order to run these on my own and see if I can get some more diagnostic information.

**Comment by maljovec** [comment]
Hmm... The only failure I get is one that was consistent when I tried the cluster tests on devel: mpiqsub_limitnode. Can I try invalidating the qsub tests? Is there a reason these tests are failing now?

**Comment by maljovec** [comment]
The failure I was seeing could have been because I did not allocate enough resources for my tests:

```
IOError: Cannot run with given parameters because no nodes have numMPI 4 available and NoSplitNode is False and LimitNode is True
```
It looks like this could be because I was on an interactive node with only one CPU requested?

**Comment by alfoa** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. ***ok***
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts). ***XSD modified***
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. ***ok***
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. ***ok***
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. ***test added***
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. ***n/a***
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. ***#122***
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added? ***n/a***

**Comment by alfoa** [comment]
The check list passed.

PR can be accepted.

**Pull request merged**